File: oculus/permissions.py
"""Permission gates before tool execution."""

from __future__ import annotations

from pathlib import Path
from typing import Any, Literal

from oculus.config import WORKDIR

Decision = Literal["allow", "deny", "ask"]

DENY_LIST = ["rm -rf /", "sudo", "shutdown", "reboot", "mkfs", "dd if="]
DESTRUCTIVE_HINTS = ["rm ", "> /etc/", "chmod 777"]


def _escapes_workspace(path: str) -> bool:
    raw = Path(path)
    if raw.is_absolute():
        try:
            return not raw.resolve().is_relative_to(WORKDIR.resolve())
        except (ValueError, OSError):
            return True
    try:
        resolved = (WORKDIR / raw).resolve()
        return not resolved.is_relative_to(WORKDIR.resolve())
    except (ValueError, OSError):
        return True


def check_permission(tool_name: str, tool_input: dict[str, Any]) -> Decision:
    if tool_name == "bash":
        command = tool_input.get("command", "")
        for pattern in DENY_LIST:
            if pattern in command:
                return "deny"
        for hint in DESTRUCTIVE_HINTS:
            if hint in command:
                return "ask"

    if tool_name in ("write_file", "edit_file") and _escapes_workspace(
        tool_input.get("path", "")
    ):
        return "ask"

    return "allow"

File: src/permissions.ts
import path from "node:path";

import { getWorkdir } from "./config.js";

export type Decision = "allow" | "deny" | "ask";

const DENY_LIST = ["rm -rf /", "sudo", "shutdown", "reboot", "mkfs", "dd if="];
const DESTRUCTIVE_HINTS = ["rm ", "> /etc/", "chmod 777"];

function escapesWorkspace(filePath: string): boolean {
  const workdir = getWorkdir();
  const raw = path.isAbsolute(filePath) ? path.resolve(filePath) : path.resolve(workdir, filePath);
  const rel = path.relative(workdir, raw);
  return rel.startsWith("..") || path.isAbsolute(rel);
}

export function checkPermission(
  toolName: string,
  toolInput: Record<string, unknown>,
): Decision {
  if (toolName === "bash") {
    const command = String(toolInput.command ?? "");
    for (const pattern of DENY_LIST) {
      if (command.includes(pattern)) {
        return "deny";
      }
    }
    for (const hint of DESTRUCTIVE_HINTS) {
      if (command.includes(hint)) {
        return "ask";
      }
    }
  }

  if (
    (toolName === "write_file" || toolName === "edit_file") &&
    escapesWorkspace(String(toolInput.path ?? ""))
  ) {
    return "ask";
  }

  return "allow";
}
